crypto/rand/drbg_lib.c

```diff
@@ -267,9 +268,10 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
         goto end;
     }
 
-    if (drbg->max_noncelen > 0 && drbg->get_nonce != NULL) {
-        noncelen = drbg->get_nonce(drbg, &nonce, drbg->strength / 2,
-                                   drbg->min_noncelen, drbg->max_noncelen);
+    if (drbg->max_noncelen > 0) {
+        if (drbg->get_nonce != NULL)
+            noncelen = drbg->get_nonce(drbg, &nonce, drbg->strength / 2,
+                                       drbg->min_noncelen, drbg->max_noncelen);
```

Comment:
You are right, the nonce is mandatory when the derivation function is used. It seems like this was indeed overlooked in the FIPS DRBG!


I would suggest a slightly different reformulation of the condition: Just move the inner `if` one level out, like so:

```
    if (drbg->max_noncelen > 0 && drbg->get_nonce != NULL)
        noncelen = drbg->get_nonce(drbg, &nonce, drbg->strength / 2,
                                   drbg->min_noncelen, drbg->max_noncelen);

    if (noncelen < drbg->min_noncelen
	    || noncelen > drbg->max_noncelen) {
        RANDerr(RAND_F_RAND_DRBG_INSTANTIATE, RAND_R_ERROR_RETRIEVING_NONCE);
        goto end;
    }
```

I think this form is equivalent to yours, but in this formulation the analogy to the `get_entropy` conditions a view lines becomes more evident:

```
    if (drbg->get_entropy != NULL)
        entropylen = drbg->get_entropy(drbg, &entropy, drbg->strength,
                                   drbg->min_entropylen, drbg->max_entropylen);
    if (entropylen < drbg->min_entropylen
        || entropylen > drbg->max_entropylen) {
        RANDerr(RAND_F_RAND_DRBG_INSTANTIATE, RAND_R_ERROR_RETRIEVING_ENTROPY);
        goto end;
    }
```

Reply:
I actually wanted to make an other check and that is that nonce != NULL, for which I guess it's easiest to have the if like I changed it.

Reply:
I still believe that my suggestion is preferrable, because it emphasizes the similiarities between the `get_nonce` and the `get_entropy` code snippet.

As for the `entropy == NULL || ` and `nonce == NULL || ` checks you added: The checks are redundand because it is by contract of the callback interface that a successful return value (`entropylen>0` resp. `noncelen > 0`) implies that `entropy != NULL` resp. `nonce != NULL` and points to a valid buffer of the respective length. So the additional check only makes the condition look more complicated.